src/bgw/job_stat.c

```diff
@@ -186,30 +188,51 @@ calculate_jitter_percent()
 /* For failures we have standard exponential backoff based on consecutive failures
  * along with a ceiling at schedule_interval * MAX_INTERVALS_BACKOFF */
 static TimestampTz
-calculate_next_start_on_failure(TimestampTz last_finish, int consecutive_failures, BgwJob *job)
+calculate_next_start_on_failure(TimestampTz finish_time, int consecutive_failures, BgwJob *job)
 {
 	float8 jitter = calculate_jitter_percent();
 	/* consecutive failures includes this failure */
 	float8 multiplier = 1 << (consecutive_failures - 1);
+	volatile TimestampTz res;
+	TimestampTz last_finish;
+	if (!IS_VALID_TIMESTAMP(finish_time))
+	{
+		elog(LOG, "calculate_next_start_on_failure, got bad finish_time");
+		last_finish = ts_timer_get_current_timestamp();
+	}
+	PG_TRY();
+	{
+		/* ival = retry_period * 2^(consecutive_failures - 1)  */
+		Datum ival = DirectFunctionCall2(interval_mul,
+										 IntervalPGetDatum(&job->fd.retry_period),
+										 Float8GetDatum(multiplier));
 
-	/* ival = retry_period * 2^(consecutive_failures - 1)  */
-	Datum ival = DirectFunctionCall2(interval_mul,
-									 IntervalPGetDatum(&job->fd.retry_period),
-									 Float8GetDatum(multiplier));
-
-	/* ival_max is the ceiling = MAX_INTERVALS_BACKOFF * schedule_interval */
-	Datum ival_max = DirectFunctionCall2(interval_mul,
-										 IntervalPGetDatum(&job->fd.schedule_interval),
-										 Float8GetDatum(MAX_INTERVALS_BACKOFF));
+		/* ival_max is the ceiling = MAX_INTERVALS_BACKOFF * schedule_interval */
+		Datum ival_max = DirectFunctionCall2(interval_mul,
+											 IntervalPGetDatum(&job->fd.schedule_interval),
+											 Float8GetDatum(MAX_INTERVALS_BACKOFF));
 
-	if (DatumGetInt32(DirectFunctionCall2(interval_cmp, ival, ival_max)) > 0)
-		ival = ival_max;
+		if (DatumGetInt32(DirectFunctionCall2(interval_cmp, ival, ival_max)) > 0)
+			ival = ival_max;
 
-	/* Add some random jitter to prevent stampeding-herds, interval will be within about +-13% */
-	ival = DirectFunctionCall2(interval_mul, ival, Float8GetDatum(1.0 + jitter));
+		/* Add some random jitter to prevent stampeding-herds, interval will be within about +-13%
+		 */
+		ival = DirectFunctionCall2(interval_mul, ival, Float8GetDatum(1.0 + jitter));
 
-	return DatumGetTimestampTz(
-		DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
+		res = DatumGetTimestampTz(
+			DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
+	}
+	PG_CATCH();
+	{
+		TimestampTz nowt;
+		FlushErrorState();
```

Comment:
This catching-and-proceeding behavior is something I've been trying to eradicate, because it is in most cases wrong. See this PostgreSQL hackers thread: https://www.postgresql.org/message-id/27190.1508727890%40sss.pgh.pa.us

I don't know if the semantics are different here if this runs in a background worker, but one serious issue here is that in the `PG_CATCH()` clause, you're in the error memory context and now it seems the execution is continuing "normal" (non error) execution on this context, which certainly would lead to bad behavior. I strongly think we should stay away from this if possible.

Otherwise, I believe (in accordance with the mailing list thread) that you must do something like:

```
        MemoryContext oldctx = CurrentMemoryContext;                                              
		BeginInternalSubTransaction("error expected");                                            
		PG_TRY();                                                                                 
		{                                                                                         
			/* do stuff that might throw error */                                              
		}                                                                                          
		PG_CATCH();                                                                                
		{                                                                                          
			/* Error handling */                                                        
			RollbackAndReleaseCurrentSubTransaction();                                             
			FlushErrorState();                                                                     
		}                                                                                          
		PG_END_TRY();                                                                             
		MemoryContextSwitchTo(oldctx); 
```

Reply:
The problem is we cannot allow the scheduler to throw errors and exit.. This stops bgw  jobs from being scheduled.This is not caused by user input but potential exceptions due to the calculation. I will modify the code to use this approach.